### ISSUE TITLE
Idea: Multiple Values for a Property

### ISSUE THREAD

**Issue by AdamSobieski** [w3c/csswg-drafts#9490]
Hello. I would like to propose an idea for CSS. As you are aware, historically, CSS properties could have one value and this value was determined utilizing an algorithm, [the cascade](https://developer.mozilla.org/en-US/docs/Web/CSS/Cascade).

What if a property could have multiple values aggregated together from multiple sources, like adding elements to a set?

Consider the following syntax
```css
div { prop: +x }
#el { prop: +y  }
.cls { prop: +z }
```
which hopes to communicate that
```html
<div id="el" class="cls" />
```
would have a set of values, `{x, y, z}`, for its property `prop`. It would obtain `x` from being a `div` element, `y` from having an id of `el`, and `z` from having a class of `cls`.

With respect to how to go about adding this functionality to CSS, we might look at and consider the CSSOM, specifically [CSSStyleDeclaration](https://drafts.csswg.org/cssom/#cssstyledeclaration) which provides the [getPropertyValue()](https://drafts.csswg.org/cssom/#dom-cssstyledeclaration-getpropertyvalue) method which returns a [`CSSOMString`](https://webidl.spec.whatwg.org/#idl-DOMString) value rather than an [`object`](https://webidl.spec.whatwg.org/#idl-object) (an `object` could be either a string or an array or list of strings). Perhaps a string representation of a set-based property's value returned by `getPropertyValue()` could be a delimited string of elements? Perhaps there could also be a `getPropertyValues()` function which would return an array or list of string-based values, this array or list having one element in it for single-value scenarios?

I hope that these ideas are of some interest to the group. What do you think of the capability for properties to have multiple values? Do any other syntax possibilities come to mind with which to express these ideas?

**Comment by Loirooriol** [comment]
See prior discussion in #1594

**Comment by AdamSobieski** [comment]
Thank you @Loirooriol.

An important keyword with respect to that [prior discussion](https://github.com/w3c/csswg-drafts/issues/1594) from 2017 is "additive".

Also, as @dbaron pointed out, @Hixie and others were discussing these topics, supplementing, prepending, and appending multiple values for CSS properties, back in 1999: https://lists.w3.org/Archives/Public/www-style/1999Oct/0025.html .

I hope that the use cases for multiple-valued CSS properties expand beyond CSS animation scenarios.

**Comment by mirisuzanne** [comment]
You show some syntax ideas, but not really a use-case. In practice, what would an author use this to achieve? What's the purpose of the feature?

**Comment by AdamSobieski** [comment]
Yes, concrete use-cases are needed for iterable and concatenable style properties.

Brainstorming, in CSS, there are different kinds of space-delimited values. Some are shorthand syntax conventions for other properties, e.g.:

```css
margin: 10px 20px 30px 40px;
```

Others are more set-like or list-like space-delimited values:

```css
transform: rotate(45deg) scale(1.5) translateX(10px);
```

One of the abstract uses of the syntax brainstorming and ideas, above, would be to enable new functionalities with respect to set-like and list-like style properties.

Interestingly, if they could be interpreted as space-delimited strings, iterable style property values could be backwards compatible. That is, `{x, y, z, w}` could also be interpreted as: `"x y z w"`.

Set-like or list-like property values, some of these presently space-delimited, could be considered for iteration and concatenation scenarios, if these capabilities were possible in CSS.

**Comment by Crissov** [comment]
In short: there is no use case!?

**Comment by AdamSobieski** [comment]
In short: I don't yet have a concrete use case example, on hand, for the expressiveness. However, someone else might!

There are three abstract topics from the initial post:

1. concatenating properties' values. 
2. iterable properties.
3. combining these ideas, concatenating iterable properties' values.

While any possible eventual syntax might vary, here are some examples of the three topics:

## Concatenating Properties' Values

```css
.cls1 { ^prop: value1 }
.cls2 { ^prop: value2 }
```
```html
<div id="el" class="cls1 cls2" />
```
```js
let element = document.getElementById('el');
let x = element.style.getPropertyValue('prop'); // == "value1 value2"
```

## Iterable Properties
```css
.cls3{ *prop: yield(x) yield(y) }
```
```html
<div id="el" class="cls3" />
```
```js
let element = document.getElementById('el');
let value = element.style.getPropertyValue('prop'); // "x y"
for (const x of element.style.getPropertyValues('prop'))
{
  // x, y
}
```

## Concatenating Iterable Properties' Values
```css
.cls4 { ^*prop: yield(x) yield(y) }
.cls5 { ^*prop: yield(z) yield(w) }
```
```html
<div id="el" class="cls4 cls5" />
```
```js
let element = document.getElementById('el');
let value = element.style.getPropertyValue('prop'); // "x y z w"
for (const x of element.style.getPropertyValues('prop'))
{
  // x, y, z, w
}
```
